Updates to manifest.json permissions and optional permissions compatibility data

rebloor opened the pull request:
Changes to:

* add a complete list of available permissions to permissions.json
* correct Firefox for Android support details for bookmarks and history permissions in optional_permissions.json

wbamberg commented on webextensions/manifest/permissions.json:
```diff
@@ -85,18 +169,147 @@
+              },
+              "firefox": {
+                "notes": [
+                  "Available as an alias to the <code>"Menus"</code> permission."
```
This is probably the line that causes the Travis failure, where the `"` needs to be escaped. Actually, it should be: `<code>menus</code>` (lowercase and not quoted).

rebloor commented:
Hi Will, there where a couple of other errors, which I've fixed. However, npm test is now giving me this:

webextensions/manifest/permissions.json
  File : /Users/sherpa/Documents/repositories/rebloor/browser-compat-data/webextensions/manifest/permissions.json
  JSON schema – 1 error(s)
   item['webextensions']['manifest']['permissions'] should NOT have additional properties
  Style – OK 
  Browser versions – OK  

I'm not seeing anything in the file that suggests I have added properties there aren't defined in the schema, such as typos, duplication etc. Any suggestions on how I find out what this is referring to?

Elchi3 commented on webextensions/manifest/permissions.json:
```diff
@@ -106,7 +319,7 @@
           }
         },
-        "clipboardRead": {
+        "downloads.open": {
```
Sorry about the error message. It's really unhelpful. The issue here is that dots aren't allowed in identifiers, as oftentimes features should rather be sub features in that case and I believe this applies here as well.

So, this should be called `open` and be nested under `downloads` like this:

```json
        "downloads": {
          "__compat": {
            "support": {
              "chrome": {
                "version_added": true
              },
              "edge": {
                "version_added": false
              },
              "firefox": {
                "version_added": "47"
              },
              "firefox_android": {
                "version_added": "48"
              },
              "opera": {
                "version_added": true
              }
            }
          },
          "open": {
            "__compat": {
              "support": {
                "chrome": {
                  "version_added": true
                },
                "edge": {
                  "version_added": false
                },
                "firefox": {
                  "version_added": "48"
                },
                "firefox_android": {
                  "version_added": "48"
                },
                "opera": {
                  "version_added": true
                }
              }
            }
          }
        },
```

rebloor replied:
Hi Florian, that resolved the errors thrown by npm test - although I still got errors when I tried to build the HTML locally using npm run render $query $dept $aggregateMode - however, the CI build doesn't seem to have thrown any errors, so hopefully we are all good. Thanks for your help.

wbamberg replied:
> So, this should be called open and be nested under downloads like this:

@Elchi3, I'm not sure about this :(. `downloads.open` is a potential value for the `permissions` key: it's a separate value from `downloads`, not a subfeature of it called `open`.

https://developer.mozilla.org/en-US/Add-ons/WebExtensions/manifest.json/permissions#API_permissions

So I'm not sure what the right thing is here. Perhaps we should call it `downloads_open` or something, and include the value in the description field? And then we should probably do the same thing for the other permissions values, to be consistent.

Elchi3 replied:
Oh, I see. Sorry about this! In this case, my comment about the nesting was wrong. 

I agree to call it downloads_open, and to add a description. So it will be like this:

```json
"downloads": {
  "__compat": {
    "support": {
      "chrome": {
        "version_added": true
      },
      "edge": {
        "version_added": false
      },
      "firefox": {
        "version_added": "47"
      },
      "firefox_android": {
        "version_added": "48"
      },
      "opera": {
        "version_added": true
      }
    }
  }
},
"downloads_open": {
  "__compat": {
    "description": "<code>downloads.open</code>",
    "support": {
      "chrome": {
        "version_added": true
      },
      "edge": {
        "version_added": false
      },
      "firefox": {
        "version_added": "48"
      },
      "firefox_android": {
        "version_added": "48"
      },
      "opera": {
        "version_added": true
      }
    }
  }
}
```

rebloor replied:
Hi, so do I need to add a description just to "download_open", or as Will seemed to suggest, to all the items?

Elchi3 replied:
To me it looks like a description for `download_open` would be sufficient for now. (just like in the json I suggested above). Will should have the final say on this, though. :-)

wbamberg replied:
> To me it looks like a description for download_open would be sufficient for now. (just like in the json I suggested above). Will should have the final say on this, though. :-)

That would work, but it feels hacky to me. If `downloads.open` isn't an identifier, why are the other permission strings? It seems more consistent to do it for all the permissions, if we're going to do it for one of them.

Elchi3 commented:
This looks good now structurally, but I would like to have @wbamberg review the actual content here.

rebloor commented:
I've committed a change with @Elchi3 suggestion, on the basis that if it is sufficient it completes this change. If I need to incorporate descriptions for all the items, I'll do that on your confirmation @wbamberg

wbamberg commented:
As said here: https://github.com/mdn/browser-compat-data/pull/928#discussion_r167096347 I think we should do this for all the permissions.

rebloor commented:
@wbamberg I've pushed another update to the permissions details, which includes descriptions added to each of the items.

rebloor commented:
Thanks @wbamberg just glad to know it all done and dusted :-)
